src/include/codegen/query_cache.h

```diff
@@ -0,0 +1,86 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// query_cache.h
+//
+// Identification: src/include/codegen/query_cache.h
+//
+// Copyright (c) 2015-17, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <list>
+#include "codegen/query.h"
+#include "common/singleton.h"
+#include "planner/abstract_plan.h"
+
+namespace peloton {
+namespace codegen {
+
+// Query cache implementation that maps an AbstractPlan with a CodeGen query
+// using LRU eviction policy. The cache is implemented as a singleton.
+// Potential enhancements (major):
+//   1) Persistency may increase the performance when rebooted
+//   2) Apply other eviction policies
+//     e.g. Keep some heavy compilation workloads by mixing policies
+//   3) Have a cache per table
+// Potential enhancements (minor):
+//   1) Manually keep some of the compiled results in the cache
+//   2) Configure the cache size
+class QueryCache : public Singleton<QueryCache> {
+ public:
+  Query* Find(const std::shared_ptr<planner::AbstractPlan> &key);
+
+  void Add(const std::shared_ptr<planner::AbstractPlan> &key,
+           std::unique_ptr<Query> &&val);
+
+  size_t GetSize() { return size_; }
+
+  void SetSize(size_t size) {
```

Comment:
Size and count are a little confusing. Can you rename `size` to capacity to indicate an upper bound?